ml/src/main/scala/ai/h2o/sparkling/ml/params/HasGamCols.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.h2o.sparkling.ml.params
+
+import ai.h2o.sparkling.H2OFrame
+import ai.h2o.sparkling.ml.algos.H2OAlgoCommonUtils
+
+trait HasGamCols extends H2OAlgoParamsBase with H2OAlgoCommonUtils {
+  protected val gamCols = nullableStringArrayParam(name = "gamCols", doc = "Predictor column names for gam.")
+
+  setDefault(gamCols -> null)
+
+  def getGamCols(): Array[String] = $(gamCols)
+
+  def setGamCols(value: Array[String]): this.type = set(gamCols, value)
+
+  override private[sparkling] def getH2OAlgorithmParams(trainingFrame: H2OFrame): Map[String, Any] = {
+    super.getH2OAlgorithmParams(trainingFrame) ++ Map("gam_columns" -> getGamCols())
+  }
+
+  override private[sparkling] def getSWtoH2OParamNameMap(): Map[String, String] = {
+    super.getSWtoH2OParamNameMap() + ("gamCols" -> "gam_columns")
+  }
+
+  override private[sparkling] def getExcludedCols(): Seq[String] = {
+    val gamCols = getGamCols()
+    if (gamCols != null) {
+      super.getExcludedCols() ++ gamCols
```

Comment:
Note: GAM columns are part of the training dataset, but are excluded from the list of feature columns. @michalkurka @wendycwong @Pscheidl is that ok?

Reply:
If a user doesn't specify feature columns. The list of feature columns is derived from all columns of the training dataset minus the label column and gam columns:
```
    val algo = new H2OGAM()
      .setSplitRatio(0.8)
      .setSeed(1)
      .setGamCols(Array("PSA", "AGE"))
      .setLabelCol("CAPSULE")
    val model = algo.fit(dataset) // [ID,CAPSULE,AGE,RACE,DPROS,DCAPS,PSA,VOL,GLEASON]
    algo.getFeaturesCols() // ->  [ID,RACE,DPROS,DCAPS,VOL,GLEASON]
```

columns of the training H2O Frame: 
```
[ID,RACE,DPROS,DCAPS,VOL,GLEASON,CAPSULE,PSA,AGE]
```

parameters sent H2O-3:
```
Map(alpha -> null, tweedie_link_power -> 0.0, gam_columns -> [Ljava.lang.String;@4eb5cdae, bs -> null, intercept -> true, non_negative -> false, stopping_rounds -> 0, ignored_columns -> null, max_categorical_levels -> 10, huber_alpha -> 0.9, class_sampling_factors -> null, prior -> -1.0, seed -> 1, tweedie_variance_power -> 0.0, categorical_encoding -> AUTO, keep_cross_validation_fold_assignment -> false, response_column -> CAPSULE, interactions -> null, fold_assignment -> AUTO, standardize -> false, scale -> null, compute_p_values -> false, lambda_search -> false, distribution -> AUTO, max_after_balance_size -> 5.0, lambda -> null, fold_column -> null, ignore_const_cols -> true, balance_classes -> false, remove_collinear_columns -> false, missing_values_handling -> MeanImputation, custom_metric_func -> null, nlambdas -> -1, max_runtime_secs -> 0.0, solver -> AUTO, score_each_iteration -> false, model_id -> null, max_active_predictors -> -1, beta_epsilon -> 1.0E-4, offset_column -> null, theta -> 0.0, custom_distribution_func -> null, link -> family_default, early_stopping -> true, weights_column -> null, parallelize_cross_validation -> true, validation_frame -> frame_rdd_251606261468_part1, export_checkpoints_dir -> null, max_iterations -> -1, objective_epsilon -> -1.0, gradient_epsilon -> -1.0, knot_ids -> null, nfolds -> 0, training_frame -> frame_rdd_251606261468_part0, beta_constraints -> null, stopping_tolerance -> 0.001, family -> AUTO, keep_cross_validation_predictions -> false, keep_gam_cols -> false, quantile_alpha -> 0.5, gainslift_bins -> -1, num_knots -> null, tweedie_power -> 1.5, stopping_metric -> AUTO, keep_cross_validation_models -> true, obj_reg -> -1.0)
```

Reply:
Another case:
If a user specify a list of features and the feature also in the list of gam columns, the feature is extracted from the list of features:
```
    val algo = new H2OGAM()
      .setSplitRatio(0.8)
      .setSeed(1)
      .setFeaturesCols("DPROS", "DCAPS", "RACE", "PSA", "GLEASON")
      .setGamCols(Array("PSA", "AGE"))
      .setLabelCol("CAPSULE")
    val model = algo.fit(dataset) // [ID,CAPSULE,AGE,RACE,DPROS,DCAPS,PSA,VOL,GLEASON]
    algo.getFeaturesCols() // ->  [DPROS,DCAPS,RACE,GLEASON]
```
columns of the training H2O Frame:
```
[DPROS,DCAPS,RACE,GLEASON,CAPSULE,PSA,AGE]
```
parameters sent H2O-3:
```
Map(alpha -> null, tweedie_link_power -> 0.0, gam_columns -> [Ljava.lang.String;@247246ea, bs -> null, intercept -> true, non_negative -> false, stopping_rounds -> 0, ignored_columns -> null, max_categorical_levels -> 10, huber_alpha -> 0.9, class_sampling_factors -> null, prior -> -1.0, seed -> 1, tweedie_variance_power -> 0.0, categorical_encoding -> AUTO, keep_cross_validation_fold_assignment -> false, response_column -> CAPSULE, interactions -> null, fold_assignment -> AUTO, standardize -> false, scale -> null, compute_p_values -> false, lambda_search -> false, distribution -> AUTO, max_after_balance_size -> 5.0, lambda -> null, fold_column -> null, ignore_const_cols -> true, balance_classes -> false, remove_collinear_columns -> false, missing_values_handling -> MeanImputation, custom_metric_func -> null, nlambdas -> -1, max_runtime_secs -> 0.0, solver -> AUTO, score_each_iteration -> false, model_id -> null, max_active_predictors -> -1, beta_epsilon -> 1.0E-4, offset_column -> null, theta -> 0.0, custom_distribution_func -> null, link -> family_default, early_stopping -> true, weights_column -> null, parallelize_cross_validation -> true, validation_frame -> frame_rdd_25838535177_part1, export_checkpoints_dir -> null, max_iterations -> -1, objective_epsilon -> -1.0, gradient_epsilon -> -1.0, knot_ids -> null, nfolds -> 0, training_frame -> frame_rdd_25838535177_part0, beta_constraints -> null, stopping_tolerance -> 0.001, family -> AUTO, keep_cross_validation_predictions -> false, keep_gam_cols -> false, quantile_alpha -> 0.5, gainslift_bins -> -1, num_knots -> null, tweedie_power -> 1.5, stopping_metric -> AUTO, keep_cross_validation_models -> true, obj_reg -> -1.0)
```

Reply:
@michalkurka @wendycwong @Pscheidl Do you think it should behave differently?

Reply:
I've change the implementation so that the gam column are independent
```
    val algo = new H2OGAM()
      .setSplitRatio(0.8)
      .setSeed(1)
      .setFeaturesCols("DPROS", "DCAPS", "RACE", "PSA", "GLEASON")
      .setGamCols(Array("PSA", "AGE"))
      .setLabelCol("CAPSULE")
    val model = algo.fit(dataset) // [ID,CAPSULE,AGE,RACE,DPROS,DCAPS,PSA,VOL,GLEASON]
    algo.getFeaturesCols() // ->  [DPROS,DCAPS,RACE,PSA,GLEASON]
```

columns of the training H2O Frame:
```[DPROS,DCAPS,RACE,PSA,GLEASON,CAPSULE,AGE]```